Rich Images

hadim opened the pull request:
- `net.imagej.notebook` code integrated (temporary)
- `net.imagej.notebook.NotebookService` has been renamed to `net.imagej.notebook.ImageJNotebookService`.

For now `ImageJNotebookService` are called manually in `NotebookService`. Ideally, I would like this to be automatic.

hadim commented:
I could put all the `NotebookService.display*()" methods in various `ScijavaService` plugins that are tracked by `NotebookService`.

ctrueden commented:
I think we should rename `imagej-notebook` to `scijava-notebook` and remove the image-specific stuff.

Then we need an extensibility mechanism for defining how to display/convert various output types to notebook-friendly form. My initial thought is a `NotebookOutput` plugin type, although these end up being very similar to `Converter` plugins. It's mostly just that we need a list of desired notebook-friendly mappings. So, maybe more clever would be for the `scijava-notebook` to simply hardcode the list of all supported output types, in some priority order, and then ask the `ConvertService` for each output object: "can you convert this output to any of the supported types on my list?" If we did that, we would not need a `NotebookOutput` plugin type probably, nor an `imagej-notebook` layer at all anymore.

hadim commented:
At the moment I am using the `org.scijava.jupyter.notebook` namespace but you're right it's maybe better to use `org.scijava.notebook`.

Everything lives in this repo for now and I'd like to keep it like that while things are not stable yet. Then I am perfectly fine to move stuff into smaller repos. Perhaps we can do that when the kernel will be integrated to the scijava org?

For the mechanism, if I get you correctly you want to define a `NotebookOutputConverter` plugin for each supported type (`String`, `List`, `Table`, `Map`, `Img`, etc). A `NotebookService` will discover those converters at runtime (or be hard-coded).

And then we could something like `nb.display(img)` or similar.

Am I correct?

hadim commented:
After some investigations, the `Converter` and `ConvertService` framework seems to be the way to go.

Input type can be whatever we want. Output type has to be a `MIMEContainer` which is a Beakerx class use to represent the output sent to the Jupyter client. Then the Jupyter client decides how to show the message according to the mimetype.

The issue is that for a given input we may want to have differents possible output and let the user or the Jupyter client choose. For example, a JSON representing a table could be printed as a JSON tree or as a table. An image could be displayed as a PNG, a JPG or with a custom widget feed directly by an `Img` stream.

My solution is to create a `NotebookOutputConverter` (similar to the Scijava `NumberToNumberConverter`) that will convert on demand a given input to a MIMEContainer.

To allow multiple outputs for a same input type, I want to create classes that inherit from `MIMEContainer`. For example : 

- `StringOutput`
- `ListOutput`
- `StringJsonOutput`
- `TreeJsonOutput`
- `TableOutput`
- `PNGImageOutput`
- `JPGImageOutput`
- `WidgetImageOutput`
- etc

Then have some glue code in the `NotebookService` to allow the user to call `notebookService.display(img)`. the `display()` method will select the first most appropriate `MIMEContainer` output type.

Helpers methods in `NotebookService` will allow alternative output types such as `notebookService.displayAsJPG(img)`, `notebookService.displayAsWidget(img)`.

hadim commented:
Here is a `NotebookOutputConverter` with a `StringToNotebookConverter` : https://github.com/hadim/scijava-jupyter-kernel/tree/432dc6324cd9676b822a2b8296f6d6494e819eea/src/main/java/org/scijava/jupyter/notebook/converter

but the following code returns `null` : 

```groovy
// @ConvertService cs
import com.twosigma.beaker.mimetype.MIMEContainer
cs.convert("test", com.twosigma.beaker.mimetype.MIMEContainer.class)
```

hadim commented:
Here is what I have implemented so far.

A `NotebookOutputConverter` which is an abstract converter. Input can be anything and output is of  `NotebookOutput` type.

Some output types all inheriting from `NotebookOutput` : 

- `PlainNotebookOutput`
- `HTMLNotebookOutput`
- `JPGNotebookOutput`
- `JSONNotebookOutput`
- `LatexNotebookOutput`
- `MarkdownNotebookOutput`
- `PNGNotebookOutput`

And the central part of the system are the converters that inherit from `NotebookOutputConverter`. For example, the converter `DatasetToPNGNotebookConverter` take a dataset as input and convert it to a PNG that the Jupyter client can display.

So it's possible to do things like that : 

```groovy
// @ImageJ ij
// @ConvertService cs
ij_logo = ij.io().open("http://imagej.net/skins/imagej-128.png")
cs.convert(ij_logo, org.scijava.jupyter.notebook.converter.ouput.PNGImageNotebookOutput.class)
```

I consider the above code very low level because the user should not worry about `PNGImageNotebookOutput` at all.

So I have implemented a `NotebookService.display(Object object)` method that automatically chooses the output type according to the input. We can also create other methods to help the user to specify what kind of output he prefers.

ctrueden commented on src/main/java/org/scijava/jupyter/notebook/DefaultNotebookService.java:
```diff
@@ -56,9 +56,30 @@
+    @Override
+    public Object display(Object object) {
+
+        if (object.getClass() == String.class) {
```
Instead of hardcoding this: why not simply type the converters to produce `NotebookOutput` generically. Then you can write `convertService.convert(object, NotebookOutput.class)` agnostically without the case logic here. As long as each kind of object has a "preferred" (i.e., highest priority) type of `NotebookOutput` that it will actually produce, this should work well. The only problem will be the case you raise where we want the user to select something; that is a separate issue though which is also not solved by the hardcoded case logic here.

hadim replied:
Good point. I'll use that. And for the user who want to select a specific output we can just create other methods such as `displayAsJPEG()`, `displayAsJSONTree()`, etc

hadim commented:
I like what we have now. The API is pretty robust and should be easily extensible. I'll merge it so we can move on.

Comments and concerns can be made is separate issues/PRs.
